proposals/text/6001-document-store-filter-rework.md

```diff
@@ -0,0 +1,197 @@
+
+In Python code:
+
+```python
+number <= 2 AND number >= 0
+```
+
+# Detailed design
+
+Filters top level must be a dictionary.
+
+There are two types of dictionaries:
+
+- Comparison
+- Logic
+
+Top level can be either be a Comparison or Logic dictionary.
+
+Comparison dictionaries must contain the keys:
+
+- `field`
+- `operator`
+- `value`
+
+Logic dictionaries have only one mandatory key:
+
+- `conditions`
+
+`conditions` key must be a list of dictionaries, either Comparison or Logic.
+
+The following key can be omitted:
+
+- `operator`
+
+`operator` values in Comparison dictionaries must be:
+
+- `==`
+- `!=`
+- `>`
+- `>=`
+- `<`
+- `<=`
+- `in`
+- `not in`
+
+`operator` values in Logic dictionaries must be:
+
+- `NOT`
+- `OR`
+- `AND`
+
+If `operator` key in Logic dictionaries is omitted we assume it's an `AND` value.
+
+---
+
+As an example this:
+
+```
+{
+    "$and": {
+        "type": {"$eq": "article"},
+        "$or": {"genre": {"$in": ["economy", "politics"]}, "publisher": {"$eq": "nytimes"}},
+        "date": {"$gte": "2015-01-01", "$lt": "2021-01-01"},
+        "rating": {"$gte": 3},
+    }
+}
+```
+
+would convert to:
+
+```
+{
+    "operator": "AND",
+    "conditions": [
+        { "field": "type", "operator": "==", "value": "article" },
+        {
+            "operator": "OR",
+            "conditions": [
+                { "field": "genre", "operator": "in", "value": ["economy", "politics"] },
+                { "field": "publisher", "operator": "==", "value": "nytimes" },
+            ]
+        },
+        { "field": "date", "operator": ">=", "value": "2015-01-01" },
+        { "field": "date", "operator": "<", "value": "2021-01-01" },
+        { "field": "rating", "operator": ">=", "value": 3 },
+    ]
+}
+```
+
+In Python code:
+
+```python
+type == "article" and (
+    genre in ["economy", "politics"] or publisher == "nytimes"
+) and date >= "2015-01-01" and date < "2021-01-01" and rating >= 3
+```
+
+Dates have been kept as strings but ideally in the new implementation they would be converted to `datetime` instances so the Document Store will be able to convert it to whatever format it needs to actually compare them. As different Document Stores might have different ways of storing the same value it's important that they handle the conversion from Python type to stored type.
+
+Another thing that in my opinion should be changed is the that filtering metadata fields must be explicitly specified for filtering. In the example above all `field`s would be prefixed with `metadata.` to get in return the expected `Document`s. e.g. `date` -> `metadata.date`
+This connects to `Document` implementation and is not the focus of this proposal, but it should be taken into account.
+
+# Drawbacks
+
+The only drawback would be that we need to adapt `MemoryDocumentStore` and `ElasticsearchDocumentStore` to use this new filtering language. These are the only two Document Stores currently created for Haystack 2.x.
```

Comment:
ChromaDB would need to be migrated as well

Reply:
Also Marqo now that I think about it.